### PR TITLE
Deck: Optionally allow to make the rerun button create a new Job

### PR DESCRIPTION
**Pull request by alvaroaleman**
With this change it is possible to set a `--rerun-creates-job` flag in Deck, it defaults to `false`. If it is set to `true` and the Deck RBAC role is extended to allow `CREATE` for `prowjobs`, the `rerun` button in Deck will actually create a new ProwJob instead of showing instructions on how to create one (Which is only helpful to someone who has access to the Prow cluster).

I've tested manually that this works, but this is is the second time I've done something with TypeScript, so it is possible that that part looks weird to someone who knows how to do it properly :grimacing: 

/assign @Katharine

### CODE REVIEW

File: prow/cmd/deck/static/prow/prow.ts [hunk 1]
```diff
@@ -652,16 +653,22 @@ function redraw(fz: FuzzySearch): void {
...
-        const copyButton = document.createElement('a');
-        copyButton.className = "mdl-button mdl-js-button mdl-button--icon";
-        copyButton.onclick = () => copyToClipboardWithToast(`kubectl create -f "${url}"`);
-        copyButton.innerHTML = "<i class='material-icons state triggered' style='color: gray'>file_copy</i>";
-        rerunElement.appendChild(copyButton);
-    };
+    var i;
+    if (rerunCreatesJob) {
+      i = icon.create("refresh", "Rerun this job");
+      i.href = `${url}`;
```
Comment (krzyzacy) [inline prow/cmd/deck/static/prow/prow.ts]: would like to see a pop up like `created prowjob $prowjobID`, or even a confirm page can help?

**Comment by krzyzacy** [comment]
cc @cjwagner @stevekuznetsov @fejta for thoughts ^^

**Comment by alvaroaleman** [comment]
> Do we want to add job-level blacklist/whitelist rerun?

If we do that should be a generic thing and not specific to triggering them via Deck. `/retest` doesn't have something like that, so IMHO adding it just in Deck is confusing.

> I think we want to log who triggered this rerun, or else it's hard to audit, even for private instances

Well, that is not possible because the auth for Deck is not integrated in Deck but instead done by some proxy that is put in front of it. If someone wants audit here, they should IMHO gather the relevant data from the proxy.

> This can work with public Prow as well, if we want to introduce github auth :-)

Yes, but that is far beyond my frontend skills :P

File: prow/cmd/deck/static/prow/prow.ts (hunk 1 above)
Comment (alvaroaleman) [reply prow/cmd/deck/static/prow/prow.ts]: Yeah, had that thought as well. Will ask one of our frontend devs for help on Monday.

**Comment by Katharine** [comment]
The ability to see the actual config used is pretty handy, as is the ability to see the prowjob ID (which is the UUID in that URL). I think that enabling this options completely removes the ability to do that. I am in favour of the functionality existing, but it would be nice to do something about the prior functionality, which went beyond a straight rerun (which is less than 5% of my use of this button).

It might also be worth somehow being clear about what "rerun" means — it uses the job configuration as of when that job was originally run (which may not be up to date), but re-resolves refs and so uses whatever the latest versions of those refs is. This may or may not be what people meant when they clicked it.

File: prow/cmd/deck/static/prow/prow.ts [hunk 2]
```diff
@@ -652,16 +653,22 @@ function redraw(fz: FuzzySearch): void {
...
-        const copyButton = document.createElement('a');
-        copyButton.className = "mdl-button mdl-js-button mdl-button--icon";
-        copyButton.onclick = () => copyToClipboardWithToast(`kubectl create -f "${url}"`);
-        copyButton.innerHTML = "<i class='material-icons state triggered' style='color: gray'>file_copy</i>";
-        rerunElement.appendChild(copyButton);
-    };
+    let i;
+    if (rerunCreatesJob) {
+      i = icon.create("refresh", "Rerun this job");
+      i.href = `${url}`;
```
Comment (Katharine) [inline prow/cmd/deck/static/prow/prow.ts]: I'd prefer to see this as an asynchronous call rather than a page reload.

**Comment by cjwagner** [comment]
> This can work with public Prow as well, if we want to introduce github auth :-)

This is something @mirandachrist will likely be working on very soon. Lets sync first thing next week and get a design doc for this so that we can implement this in an extensible way.

> The ability to see the actual config used is pretty handy

+100 It would be great if we exposed the ProwJob yaml via a dedicated button instead of providing that info indirectly and non-obviously with the rerun button.

> It might also be worth somehow being clear about what "rerun" means — it uses the job configuration as of when that job was originally run (which may not be up to date), but re-resolves refs and so uses whatever the latest versions of those refs is. This may or may not be what people meant when they clicked it.

Also a great point. It would be nice to provide both options with a UI switch. It would also be nice to have an option that runs the job newest configuration for the job instead of the spec that ran previously (we could add that in a follow up PR).

**Comment by stevekuznetsov** [comment]
I think a form fill guarded by GitHub Auth would be a much more forward-looking approach to this. +1 to Cole's comments

**Comment by alvaroaleman** [comment]
> The ability to see the actual config used is pretty handy, as is the ability to see the prowjob ID (which is the UUID in that URL). I think that enabling this options completely removes the ability to do that. I am in favour of the functionality existing, but it would be nice to do something about the prior functionality, which went beyond a straight rerun (which is in practice almost none of my use of this button — I mostly just want a quick way to see what the job was).

Woho, I didn't realize. I was actually thinking about filing an issue so `spyglass` shows some infos like the job id and the name of the pod in its job view to easen debugging, but felt like I am complaining too much already :grimacing: 

> It might also be worth somehow being clear about what "rerun" means — it uses the job configuration as of when that job was originally run (which may not be up to date), but re-resolves refs and so uses whatever the latest versions of those refs is. This may or may not be what people meant when they clicked it.

That does match with at least what I expected, although not sure if it also matches general expectations.

> I think a form fill guarded by GitHub Auth would be a much more forward-looking approach to this. +1 to Cole's comments

Completely agree. But I would like to have this in the meantime as an intermediate solution. Currently the only way to restart periodics or postsubmits is by using what yaml from Deck and I would like to make this a bit better.

File: prow/cmd/deck/static/prow/prow.ts (hunk 2 above)
Comment (alvaroaleman) [reply prow/cmd/deck/static/prow/prow.ts]: But the page has to be reloaded anyways to see the new job or is there some other way? (Feel free to point my some doc, this is far from my area of expertise)

**Comment by stevekuznetsov** [comment]
> That does match with at least what I expected, although not sure if it also matches general expectations.

I think we would want to make this an active choice from the user.

> Completely agree. But I would like to have this in the meantime as an intermediate solution. Currently the only way to restart periodics or postsubmits is by using what yaml from Deck and I would like to make this a bit better.

While I appreciate that I do want to make sure we take the time to make this clean and not err too much on the side of quick, organic changes that complicate our landscape.

**Comment by fejta** [comment]
For presubmits we need to validate that the requesting user has the ability to call `/test foo` which might not always be the case: aka someone writes `/test foo` without leaving an `/ok-to-test`

**Comment by stevekuznetsov** [comment]
Yeah, and all of those questions are much more easily solved with a web page behind github auth. I think Alva means for this PR to be an admin-level flag on less locked-down Prow instances?

**Comment by mirandachrist** [comment]
I'm working on adding a button that links to the prowjob YAML, so the config will still be viewable after the rerun button is changed.

**Comment by alvaroaleman** [comment]
> I think Alva means for this PR to be an admin-level flag on less locked-down Prow instances

Yes, exactly that, this should never get activated on a public Prow instance and that is also what the description of the flag says.

Discussed this with @cjwagner and @mirandachrist . Miranda will add another button for showing the yaml and improve what I added here by adding authorization and the possibility to change the refs.

**Comment by alvaroaleman** [comment]
So the "job config cant be displayed anymore" is not valid anymore, thanks to the dedicated button @mirandachrist added in 26eecf994b9e9f3657bb028d6534f89ea0830fd9.

@Katharine @cjwagner @stevekuznetsov Is there currently still something open you want to see as part of this PR? While the functionality is not really usable for public prow instances right now, its feature gated and shouldn't be too hard to extend with authn and/or authz.

**Comment by Katharine** [comment]
So I have one major concern about this - there is no CSRF ([Cross-Site Request Forgery](https://en.wikipedia.org/wiki/Cross-site_request_forgery)) protection present. The upshot is that a malicious user can get an authorised user to rerun a job by convincing them to follow a link. In theory, this is even a problem for an internal instance - in practice, given that you need a valid prowjob UUID to do anything, for a truly internal instance the risk is probably minimal. Externally, though, this would be a non-starter without additional CSRF protection, even once authorisation was added.

The general approach to CSRF protection is to issue users with a token on page load which is then validated when a protected action is taken. Usually this involves maintaining state, but if you can manage user/session identifiers somehow then you can pull it off using crypto instead.

Given the context of this PR, I'm not going to block on this - as long as only authorised users can discover valid prowjob UUIDs, I think the risk is pretty minimal. However, we _must_ not blindly add authentication to this and open it up more broadly. I would like to see a comment somewhere pertinent in the code reminding people of this.

I also have a second concern: the request to `/rerun` should be required to be a `POST` request, for semantic reasons. This does not meaningfully improve security, but does avoid e.g. accidentally prefetching every `/rerun` on the page and thereby rerunning all the jobs. To do this from the page, you can either arrange for JavaScript to submit a form to this end, or make the rerun buttons be submit buttons for a tiny form with a hidden field.

**Comment by alvaroaleman** [comment]
@Katharine This now does a POST request and has a warning regarding the CSRF

File: prow/cmd/deck/static/prow/prow.ts [hunk 3]
```diff
@@ -653,16 +654,28 @@ function redraw(fz: FuzzySearch): void {
 function createRerunCell(modal: HTMLElement, rerunElement: HTMLElement, prowjob: string): HTMLTableDataCellElement {
     const url = `https://${window.location.hostname}/rerun?prowjob=${prowjob}`;
     const c = document.createElement("td");
-    const i = icon.create("refresh", "Show instructions for rerunning this job");
-    i.onclick = () => {
+    let i;
+    if (rerunCreatesJob) {
+      i = icon.create("refresh", "Rerun this job");
+      i.onclick = () => {
```
Comment (mirandachrist) [inline prow/cmd/deck/static/prow/prow.ts]: Just a heads up - some people have said they'd still like to see the "kubectl create ..." popup, even if directly triggering the rerun is enabled. In my [PR](https://github.com/kubernetes/test-infra/pull/13008), I'm adding a "run" button within that popup that will actually rerun the job, so it's still displayed:
image

Comment (alvaroaleman) [reply prow/cmd/deck/static/prow/prow.ts]: @mirandachrist If you just hadn't explained that, I wouldn't have derived from the view on the screenshot that the run button runs the job tbh (But that discussion is probably better put on your PR)

**Comment by Katharine** [comment]
This looks almost good to me, but I would like to see `handleRerun` actually check the request method and only do the rerun if it is actually POST.

I'd ideally also like to see it take the prowjob name from a POSTed form rather than from the query string, but I'm not going to be picky about that one this time around.

**Comment by cjwagner** [comment]
> Given the context of this PR, I'm not going to block on this - as long as only authorised users can discover valid prowjob UUIDs, I think the risk is pretty minimal.

Anyone can discover valid ProwJob UUIDs if they can access the front end. The `/data.js`, `/prowjobs.js`, and ProwJob YAML buttons all expose valid UUIDs.
IIUC this PR is only considered safe because the feature is meant for private Prow instances that have an oauth proxy in front of `deck`?

**Comment by Katharine** [comment]
> IIUC this PR is only considered safe because the feature is meant for private Prow instances that have an oauth proxy in front of deck?

Correct. We cannot use this as part of the OAuth-ed rerun setup without further work (and there is now a comment in the relevant code reminding us of this).

**Comment by alvaroaleman** [comment]
> This looks almost good to me, but I would like to see handleRerun actually check the request method and only do the rerun if it is actually POST.

@Katharine done

**Comment by alvaroaleman** [comment]
Or did you mean for it to always check for POST, not only in the case where the rerun is actually executed?

**Comment by Katharine** [comment]
> Or did you mean for it to always check for POST, not only in the case where the rerun is actually executed?

Only in the case where the rerun is actually executed.

File: prow/cmd/deck/main.go [hunk 4]
```diff
@@ -1163,8 +1165,26 @@ func handleRerun(prowJobClient prowv1.ProwJobInterface) http.HandlerFunc {
...
-		b, err := yaml.Marshal(&pjutil)
+		newPJ := pjutil.NewProwJob(pj.Spec, pj.ObjectMeta.Labels)
+		// Be very careful about this on publicly accessible Prow instances. Even after we have authentication
+		// for the handler, we need CSRF protection.
+		// On Prow instances that require auth even for viewing Deck this is okayish, because the Prowjob UUID
+		// is hard to guess
+		// Ref: https://github.com/kubernetes/test-infra/pull/12827#issuecomment-502850414
+		if createProwJob {
+			if r.Method != http.MethodPost {
+				http.Error(w, "request must be of type POST", http.StatusBadRequest)
```
Comment (Katharine) [inline prow/cmd/deck/main.go]: Nit:

```suggestion
				http.Error(w, "request must be of type POST", http.StatusMethodNotAllowed)
```

**Pull request merged**